src/net/ifaddrs/mod.rs

```diff
@@ -0,0 +1,273 @@
+
+mod tests;
+
+pub type InterfaceMap<'a> = HashMap<String, Vec<InterfaceAddr<'a>>>;
+
+/// Represents a handle into the operating system's knowledge about network
+/// interfaces present on the system. Allows the user to iterate over
+/// interface configurations.
+pub struct InterfaceAddrs<'a> {
+    inner: *mut libc::ifaddrs,
+    current: Option<&'a libc::ifaddrs>,
+    do_free: bool,
+}
+
+impl<'a> InterfaceAddrs<'a> {
+    /// Creates an `InterfaceAddrs` from a raw pointer, without calling into
+    /// the `libc`.
+    ///
+    /// The destructor will not attempt to free memory on an InterfaceAddrs
+    /// created in this way.
+    ///
+    /// # Unsafety
+    /// The caller is responsible for making sure the given pointer is not
+    /// in invalid memory.
+    ///
+    /// # Errors
+    /// `Err(())` will be returned if `p` was void.
+    pub unsafe fn from_raw(p: *mut libc::ifaddrs) -> ::std::result::Result<InterfaceAddrs<'a>, ()> {
+        match p.as_ref() {
+            Some(r) => Ok(Self {
+                inner: p,
+                current: Some(r),
+                do_free: false,
+            }),
+            None => Err(()),
+        }
+    }
+
+    /// Produce an `InterfaceAddrs` from the system's information.
+    pub fn getifaddrs() -> Result<Self> {
+        let mut p = null_mut();
+
+        unsafe {
+            libc::getifaddrs(&mut p);
+        }
+
+        // UNSAFETY: *mut -> &'static mut. This is known to be either in valid
+        // memory or null based on the guarantees of getifaddrs()
+        match unsafe { p.as_ref() } {
+            Some(r) => Ok(Self {
+                inner: p,
+                current: Some(r),
+                do_free: true,
+            }),
+
+            None => Err(Error::from(Errno::from_i32(errno()))),
+        }
+    }
+}
+
+impl<'a> From<InterfaceAddrs<'a>> for HashMap<String, Vec<InterfaceAddr<'a>>> {
+    /// Collect an `InterfaceAddrs` into a `HashMap<String, InterfaceAddr>`.
+    fn from(ia: InterfaceAddrs<'a>) -> HashMap<String, Vec<InterfaceAddr<'a>>> {
+        let mut m = HashMap::new();
+        for i in ia {
+            if !m.contains_key(&i.name) {
+                m.insert(i.name.clone(), Vec::new());
+            }
+            // Unwrap here because contains is checked above
+            m.get_mut(&i.name).unwrap().push(i);
+        }
+
+        m
+    }
+}
+
+impl<'a> Drop for InterfaceAddrs<'a> {
+    fn drop(&mut self) {
+        if self.do_free {
+            // UNSAFETY: Calling libc FFI function which frees previously allocated
+            // memory.
+            unsafe {
+                // Ask the libc to drop free the memory it allocated when
+                // the struct was created.
+                libc::freeifaddrs(self.inner as *mut libc::ifaddrs);
+            }
+        }
+    }
+}
+
+
+/// Represents the configuration and state of a network interface.
+/// Interfaces are uniquely identified by name, and each interface is likely
```

Comment:
Needs a blank line above for the generated docs to look pretty (run rustdoc to see what I'm talking about).